File: src/localization/scripts/position_sender.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse
import time
import numpy as np
import matplotlib.pyplot as plt
from rclpy.executors import MultiThreadedExecutor
from rclpy.executors import ExternalShutdownException
from rclpy.callback_groups import ReentrantCallbackGroup

# Import des messages
from cdf_msgs.action import MoveTo
from cdf_msgs.srv import PosTrigger
from geometry_msgs.msg import Point
from cdf_msgs.msg import PositionMap
import threading


class PositionSender(Node):
    def __init__(self):
        super().__init__("position_sender_node")
        self._init_parameters()
        self._init_publishers()
        self._init_subscribers()
        self._init_clients()
        self.get_logger().info(f"Position sender node initialized.")

    def _init_parameters(self) -> None:
        self.declare_parameters(
            namespace="",
            parameters=[
                ("robot_names", rclpy.Parameter.Type.STRING_ARRAY),
                ("real_position_topic", rclpy.Parameter.Type.STRING),
                ("trigger_position_srv", rclpy.Parameter.Type.STRING),
                ("update_period", rclpy.Parameter.Type.DOUBLE),
                ("update_position_topic", rclpy.Parameter.Type.STRING),
            ],
        )
        
        self.robot_names = (
            self.get_parameter("robot_names").get_parameter_value().string_array_value
        )
        self.update_period = (
            self.get_parameter("update_period").get_parameter_value().double_value
        )
        self.current_pos = {name: None for name in self.robot_names}

    def _init_publishers(self) -> None:
        self.update_position_topic = (
            self.get_parameter("update_position_topic").get_parameter_value().string_value
        )
        self.pub_update_position = {name: self.create_publisher(PositionMap, name + '/' +  self.update_position_topic, 10) for name in self.robot_names}

    def _init_subscribers(self):
        ##Inversion pb here, Weird 
        self.real_position_topic = (
            self.get_parameter("real_position_topic").get_parameter_value().string_value
        )

        #### WTF INIT LIKE THAT WORKS, BUT USING FOR LOOP DOES NOT !?????? 
        self.create_subscription(Point, '/main_robot/' + self.real_position_topic, lambda msg: self._update_pos(msg, "main_robot"), 10)
        self.create_subscription(Point, '/ennemi_robot/' + self.real_position_topic, lambda msg: self._update_pos(msg, "ennemi_robot"), 10)

        # for name in self.robot_names:
        #     self.get_logger().info(f"{name}: {'/' + name + '/' + self.real_position_topic}")
        #     sub_init = self.create_subscription(Point, '/' + name + '/' + self.real_position_topic, lambda msg: self._update_pos(msg, name), 10)
        #     sub_init
        self.create_timer(self.update_period, self._publish_position)

    def _init_clients(self):
        self.trigger_position_srv = (
            self.get_parameter("trigger_position_srv").get_parameter_value().string_value
        )
        for name in self.robot_names:
            cli_trigger_position = self.create_client(PosTrigger, name + '/' + self.trigger_position_srv)
            while not cli_trigger_position.wait_for_service(timeout_sec=1.0):
                self.get_logger().warn(f'No position to update currently for {name}...')
            req = PosTrigger.Request()
            future = cli_trigger_position.call_async(req)
            rclpy.spin_until_future_complete(self, future)
            self.current_pos[name] = future.result().pos

    def _update_pos(self, msg, name):
        self.get_logger().info(f"update of {name}")
        self.current_pos[name] = msg

    def _publish_position(self):
        for name in self.robot_names:
            self.get_logger().info(f"name: {name}")
            if self.current_pos[name] is not None:
                msg = PositionMap()
                msg.robot = self.current_pos[name]
                self.get_logger().info(f"name: {self.current_pos[name]}")
                for n in [na for na in self.robot_names if na != name]:
                    msg.ennemis.append(self.current_pos[n])
                self.pub_update_position[name].publish(msg)

def main(args=None):
    rclpy.init(args=args)
    position_sender_node = PositionSender()
    try:
        rclpy.spin(position_sender_node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        position_sender_node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: localization/src/position_sender.cpp
#include "localization/position_sender.hpp"

#include <chrono>
#include <functional>

using namespace std::chrono_literals;

PositionSender::PositionSender() : Node("position_sender_node")
{
  init_parameters();
  init_publishers();
  init_subscribers();
  init_clients();
  RCLCPP_INFO(this->get_logger(), "Position sender node initialized.");
}

void PositionSender::init_parameters()
{
  // Declare parameters with default values.
  // You can adjust the defaults as necessary.
  this->declare_parameter<std::vector<std::string>>("robot_names", std::vector<std::string>{});
  this->declare_parameter<std::string>("real_position_topic", "");
  this->declare_parameter<std::string>("trigger_position_srv", "");
  this->declare_parameter<double>("update_period", 0.1);
  this->declare_parameter<std::string>("update_position_topic", "");

  // Get parameter values
  this->get_parameter("robot_names", robot_names_);
  this->get_parameter("real_position_topic", real_position_topic_);
  this->get_parameter("trigger_position_srv", trigger_position_srv_);
  this->get_parameter("update_period", update_period_);
  this->get_parameter("update_position_topic", update_position_topic_);

  // Initialize the current positions to nullptr for each robot.
  for (const auto & name : robot_names_) {
    current_pos_[name] = nullptr;
  }
}

void PositionSender::init_publishers()
{
  // Create a publisher for each robot.
  // The topic name is constructed as: "<robot_name>/<update_position_topic>"
  for (const auto & name : robot_names_) {
    std::string topic_name = name + "/" + update_position_topic_;
    pub_update_position_[name] = this->create_publisher<cdf_msgs::msg::PositionMap>(topic_name, 10);
  }
}

void PositionSender::init_subscribers()
{
  // Create a subscription for each robot.
  // The subscription listens on: "/<robot_name>/<real_position_topic>"
  for (const auto & name : robot_names_) {
    std::string topic_name = "/" + name + "/" + real_position_topic_;
    auto sub = this->create_subscription<geometry_msgs::msg::Point>(
      topic_name,
      10,
      [this, name](const geometry_msgs::msg::Point::SharedPtr msg) {
        this->update_pos_callback(msg, name);
      }
    );
    subscriptions_.push_back(sub);
  }

  // Create a timer to periodically call publish_position_callback.
  timer_ = this->create_wall_timer(
    std::chrono::duration<double>(update_period_),
    std::bind(&PositionSender::publish_position_callback, this)
  );
}

void PositionSender::init_clients()
{
  // For each robot, create a service client to request its initial position.
  for (const auto & name : robot_names_) {
    std::string service_name = name + "/" + trigger_position_srv_;
    auto client = this->create_client<cdf_msgs::srv::PosTrigger>(service_name);

    // Wait for the service to become available (timeout 1 second per attempt)
    while (!client->wait_for_service(1s)) {
      RCLCPP_WARN(this->get_logger(), "No position to update currently for %s...", name.c_str());
    }

    // Create the request and call the service asynchronously.
    auto request = std::make_shared<cdf_msgs::srv::PosTrigger::Request>();
    auto future_result = client->async_send_request(request);

    // Wait for the result synchronously.
    if (rclcpp::spin_until_future_complete(this->get_node_base_interface(), future_result) ==
        rclcpp::FutureReturnCode::SUCCESS)
    {
      // Save the received position.
      current_pos_[name] = std::make_shared<geometry_msgs::msg::Point>(future_result.get()->pos);
    } else {
      RCLCPP_ERROR(this->get_logger(), "Failed to call service for %s", name.c_str());
    }
  }
}

void PositionSender::update_pos_callback(const geometry_msgs::msg::Point::SharedPtr msg, const std::string & robot_name)
{
  // Update the position for the given robot.
  current_pos_[robot_name] = msg;
}

void PositionSender::publish_position_callback()
{
  // For each robot, if its current position is set, publish a PositionMap message.
  for (const auto & name : robot_names_) {
    if (current_pos_[name] != nullptr) {
      cdf_msgs::msg::PositionMap msg;
      // Set the robot's own position.
      msg.robot = *current_pos_[name];

      // For every other robot, append its position to the ennemis list.
      for (const auto & other_name : robot_names_) {
        if (other_name != name) {
          if (current_pos_[other_name] != nullptr) {
            msg.ennemis.push_back(*current_pos_[other_name]);
          } else {
            // If no position is available, push a default-constructed Point.
            msg.ennemis.push_back(geometry_msgs::msg::Point());
          }
        }
      }
      pub_update_position_[name]->publish(msg);
    }
  }
}

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<PositionSender>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
